chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md

```diff
@@ -0,0 +1,183 @@
+
+## Scaling up to improve performance
+
+The most obvious goal of scaling brokers is to unlock additional processing performance.
+While vertical scaling is also a great option, this can hit limits imposed by your infrastructure provider.
+For example, some machine types may offer great CPU performance but are severely limited in IOPS.
+Additionally, vertical scaling is often more expensive than horizontal scaling.
+It also comes with increased risk when a single machine fails because the remaining machines may already run at their limits and will then struggle to handle the additional load during failover.
+
+To show how broker scaling can improve processing performance, we reused the same cluster setup as before.
+We have 3 brokers, 6 partitions and replication factor 3.
+
+The brokers are limited to 1.35 CPUs and 4GiB RAM each. 
+They run with additional safety checks that are usually disabled in production and that slightly decrease the baseline processing performance.
+Each broker uses a small 32GiB SSD for storage, limiting them to a few thousand IOPS.
+
+We changed the processing load slightly to simulate a more realistic scenario.
+The new process model consists of 10 tasks with two timers in-between, each delaying the process instance by 1 minute.
+
+![](./ten_tasks.png)
+
+The processing load is generated by our own [benchmarking application](https://github.com/camunda/zeebe/tree/9e723b21b0e408fc2b97fd7d3f6b092af8e62dbe/benchmarks), initially starting 40 process instances per second.
+
+This results in 400 jobs created and completed per second.
+
+This stresses the 3 brokers and we see backpressure on all partitions.
+![](./perf_initial_backpressure.png)
+
+We also see a few jobs timing out, indicating that the cluster is unable to handle this load consistenly:
+![](./perf_initial_timeouts.png)
+
+We also see that many jobs are active for much longer than 1 second, even though the workers could complete them immediately.
+![](./perf_initial_job_lifetime.png)
+
+As hinted at before, much of this performance limit can be attributed to the limited IOPS of the small SSDs.
+We see this in a very high commit and write latency, while the IOPS remain stable, right at the limit of what the SSDs can handle.
+
+![](./perf_initial_iops.png)
+![](./perf_initial_commit_latency.png)
+![](./perf_initial_write_latency.png)
+
+### Expected
+
+When we scale up to 6 brokers, and thus distribute the partitions such that each broker is only leader for 1 instead of 2 partitions, we expect that processing performance improves.
+
+As these things usually go, we don't expect a doubling in performance but aiming for a 1.5x improvement seems reasonable.
+
+### Actual
+
+Shortly after scaling up and after partition leadership has balanced, we see a significant improvement in backpressure.
+![](./perf_after_backpressure.png)
+
+The job lifetime decreases dramatically, with most jobs now taking < 5ms from creation until completion.
```

Comment:
❓ Isn;t this 50 ms? 🤔

Reply:
Ah damn, I _thought_ that 5ms is surprisingly good but didn't check it. Of course 50ms is correct :+1: